<a href="https://colab.research.google.com/github/AkiraTerao/AT/blob/master/mie_scatter.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import numpy as np
import matplotlib.pyplot as pyllot
%matplotlib inline

function musgp = getMieScatter(lambda, dia, fv, npar,nmed)

% function musgp = getMieScatter(lambda, dia, fv)

%  fv            = volume fraction of spheres in medium (eg., fv = 0.05)

%  lambda        = wavelength in um (eg., lambda = 0.633)

%  dia           = sphere diameter in um (eg., dia_um = 0.0500)

%  npar          = particle refractive index (eg. polystyrene = 1.57)

%  nmed          = medium refractive index (eg., water = 1.33)

%                  Note: npar and nmed can be imaginary numbers.

%  returns musgp = [mus g musp]  

%       mus      = scattering coefficient [cm^-1]

%       g        = anisotropy of scattering [dimensionless]

%       musp     = reduced scattering coefficient [cm^-1]

%  Uses

%       Mie.m, which uses mie_abcd.m, from Maetzler 2002

In [ ]:
fv=0.05
lam=0.633 # um
dia=0.05 # um
npar=1.57 
nmed=1.33

In [ ]:
Vsphere = 4/3*np.pi*(dia/2)**3;     # volume of sphere
rho     = fv/Vsphere;           # #/um^3, concentration of spheres
m = npar/nmed;                  # ratio of refractive indices
x = np.pi*dia/(lam/nmed);       # ratio circumference/wavelength in medium

In [ ]:
if x==0: # To avoid a singularity at x=0
  result=np.array([np.real(m),np.imag(m), 0, 0, 0, 0, 0, 0, 1.5])
elif x>0: # This is the normal situation
  nmax=round(2+x+4*x**(1/3));

In [ ]:
n1=nmax-1;
n=np.arange(1,nmax+1);cn=2*n+1; c1n=n*(n+2)/(n+1); c2n=cn/n/(n+1);
x2=x*x;

mie_abcd

In [ ]:
nmax=round(2+x+4*x**(1/3));
n=np.arange(1,nmax+1); nu =(n+0.5); z=m*x; m2=m*m;
sqx= np.sqrt(0.5*np.pi/x); sqz= np.sqrt(0.5*np.pi/z);

In [ ]:
from scipy.special import jv, yv
bx = jv(nu, x)*sqx
bz = jv(nu, z)*sqz
yx = yv(nu, x)*sqx
hx = bx+1j*yx;

In [ ]:
b1x=np.concatenate([[np.sin(x)/x],bx[0:nmax-1]])
b1z=np.concatenate([[np.sin(z)/z],bz[0:nmax-1]])
y1x=np.concatenate([[-np.cos(x)/x],yx[0:nmax-1]])
h1x= b1x+1j*y1x;

In [ ]:
ax = x*b1x-n*bx;
az = z*b1z-n*bz;
ahx= x*h1x-n*hx;

In [ ]:
an = (m2*bz*ax-bx*az)/(m2*bz*ahx-hx*az);
bn = (bz*ax-bx*az)/(bz*ahx-hx*az);
cnf = (bx*ahx-hx*ax)/(bz*ahx-hx*az);
dn = m*(bx*ahx-hx*ax)/(m2*bz*ahx-hx*az);
result=np.array([an, bn, cnf, dn]);
f=result

end of mie_abcd

In [ ]:
anp=np.real(f[0,:]); anpp=np.imag(f[0,:]);
bnp=np.real(f[1,:]); bnpp=np.imag(f[1,:]);

In [ ]:
g1=np.zeros([4,nmax]); g1[0,:n1]=anp[1:nmax]; g1[1,:n1]=anpp[1:nmax]; g1[2,:n1]=bnp[1:nmax]; g1[3,:n1]=bnpp[1:nmax]

In [ ]:
dn=cn*(anp+bnp);
q=sum(dn);
qext=2*q/x2;
en=cn*(anp*anp+anpp*anpp+bnp*bnp+bnpp*bnpp);
q=sum(en);
qsca=2*q/x2;
qabs=qext-qsca;
fn=(f[0,:]-f[1,:])*cn;
gn=(-1)**n;
f[2,:]=fn*gn;
q=np.sum(f[2,:]);

In [ ]:
fn

array([2.26106741e-05-8.13505248e-03j, 1.53375253e-09-8.70456490e-05j,
       1.85852459e-14-3.59413064e-07j, 6.70441257e-20-7.74977943e-10j,
       9.52359539e-26-1.02183534e-12j])

In [ ]:
qb=q*q/x2;
asy1=c1n*(anp*g1[0,:]+anpp*g1[1,:]+bnp*g1[2,:]+bnpp*g1[3,:]);
asy2=c2n*(anp*bnp+anpp*bnpp);
asy=4/x2*np.sum(asy1+asy2)/qsca;
qratio=qb/qsca;
result=np.array([np.real(m), np.imag(m), x, qext, qsca, qabs, qb, asy, qratio]);

In [ ]:
u=result
qsca = u[4];                    # scattering efficiency, Qsca
g    = u[7];                    # anisotropy, g

A       = np.pi*dia**2/4;          # geometrical cross-sectional area, um^2
sigma_s = qsca*A;               # scattering cross-section, um^2
mus     = sigma_s*rho*1e4;      # scattering coeff. cm^-1
musp    = mus*(1-g);            # reduced scattering coeff. cm^-1


In [ ]:
print('----- choice:')
print(f'lambda = {lam: .4f} um')
print(f'diameter = {dia: .4f} um')
print(f'rho =  {rho: .4f} 1/um^3')
print(f'npar = {npar: .3f}')
print(f'nmed = {nmed: .3f}')

----- choice:
lambda =  0.6330 um
diameter =  0.0500 um
rho =   763.9437 1/um^3
npar =  1.570
nmed =  1.330


In [ ]:
 print('----- result:')
print('real(m):' , u[0])
print('imag(m):', u[1])
print('x: ', u[2])
print('qext: ', u[3])
print('qsca: ', u[4])
print('qabs: ', u[5])
print('qb: ', u[6])
print('asy: ', u[7])
print('qratio: ', u[8])
print('----- optical properties:')
print('mus =' +str(mus)+ ' cm^-1 ')
print('g = ',g)
print('musp = ' +str(mus)+ 'cm^-1')

----- result:
real(m): (1.1804511278195489+0j)
imag(m): 0j
x:  (0.33004093438186616+0j)
qext:  (0.00041530714674885413+0j)
qsca:  (0.000415307146748854+0j)
qabs:  (1.0842021724855044e-19+0j)
qb:  (-0.0005946704175842089-3.3410741837797736e-06j)
asy:  (0.018706478310797883+0j)
qratio:  (-1.4318810120159575-0.008044827087457273j)
----- optical properties:
mus =(6.2296072012328105+0j) cm^-1 
g =  (0.018706478310797883+0j)
musp = (6.2296072012328105+0j)cm^-1


In [ ]:
print(f'lambda = {lam: .4f} um')

lambda =  0.6330 um


----- result:
real(m): (1.1804511278195489+0j)

imag(m): 0j

x:  (0.33004093438186616+0j)

qext:  (0.00041530714674885413+0j)

qsca:  (0.000415307146748854+0j)

qabs:  (1.0842021724855044e-19+0j)

qb:  (-0.0005946704175842089-3.3410741837797736e-06j)

asy:  (0.018706478310797883+0j)

qratio:  (-1.4318810120159575-0.008044827087457273j)

----- optical properties:

mus =(6.2296072012328105+0j) cm^-1 

g =  (0.018706478310797883+0j)

musp = (6.2296072012328105+0j)cm^-1
